src/util/compat.hpp

```diff
@@ -34,4 +34,27 @@ inline void compiler_barrier_if_xlc_leq(unsigned ver) {
 template <typename X>
 inline constexpr bool isinf(X x) { return std::isinf(x); }
 
+// Work around a bad inlining-related optimization with icpc 16.0.3 and -xMIC-AVX512,
+// by forcing a computation.
+
+template <typename X>
+inline void sink(const X& x) {
```

Comment:
What does work is in fact to place an explicit barrier after the `make_partition`. I'll resubmit without the sink, and instead with a `compiler_barrier_if_icc_leq` which is directly akin to `compiler_barrier_if_xlc_leq`.